### PR TITLE
extract codegen into macro-rules macros

### PR DESCRIPTION
**Pull request by nikomatsakis**
This applies a pattern that I've found to be very nice to work with, in which the duchess crate has two helper crates:

* `duchess-macro-rules`, internal `macro_rules!` macro that handle most of the codegen
* `duchess-macros`, procedural macros which generate calls to those macros after doing validation and pre-processing

This combo lets each kind of macro manage the things it is good at: procedural macros for generating errors and doing some grody stuff, but macro-rules for generating the actual code.

Also, as a driveby change, adds a `dev-container.json` file which makes it easier to manage the setup for testing and developing.

r? @rcoh

### CODE REVIEW

File: macro-rules/src/java_types.rs [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+#[macro_export]
+macro_rules! argument_impl_trait {
+    ($scalar:ident $(+ $lt:lifetime)?) => {
+        impl duchess::IntoScalar< duchess::plumbing::rust_ty!($scalar) > $(+ $lt)?
+    };
+
+    ($r:tt $(+ $lt:lifetime)?) => {
+        impl duchess::IntoJava< duchess::plumbing::rust_ty!($r) > $(+ $lt)?
+    };
+}
```
Comment (rcoh) [inline macro-rules/src/java_types.rs:25]: for the macros that aren't single-pattern and pretty obvious, I think some usage examples would probably be helpful

File: macro-rules/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+mod java_types;
```
Comment (rcoh) [inline macro-rules/src/lib.rs:1]: as a small point, it's hard to find where the macros are actually defined because they all get exported into the root namespace. I guess there's no way around that though :-/

File: macro-rules/src/util.rs [hunk 3]
```diff
@@ -0,0 +1,25 @@
+#[macro_export]
+macro_rules! macro_if {
+    (if [] { $($then:tt)* }) => {};
+
+    (if false { $($then:tt)* }) => {};
+
+    (if [$($input:tt)+] { $($then:tt)* }) => {
+        $($then)*
+    };
```
Comment (rcoh) [inline macro-rules/src/util.rs:9]: I'm kinda confused, but this looks like it ignores the $input arguments? AFAICT it expands identically to `if true`

File: macro-rules/src/util.rs [hunk 4]
```diff
@@ -0,0 +1,25 @@
+#[macro_export]
```
Comment (rcoh) [inline macro-rules/src/util.rs:1]: some usage examples would be very helpful here

File: duchess-reflect/src/codegen.rs [hunk 5]
```diff
@@ -134,238 +129,30 @@ impl ClassInfo {
...
-
-            #[allow(non_camel_case_types)]
-            pub struct #struct_name<#(#java_class_generics_with_defaults,)*> {
-                _empty: std::convert::Infallible,
-                _dummy: ::core::marker::PhantomData<(#(#java_class_generics,)*)>
+        let mro_tys = self.mro(upcasts)?;
+
+        let output = quote! {
+            duchess::plumbing::setup_class! {
+                struct_name: [#struct_name],
```
Comment (rcoh) [inline duchess-reflect/src/codegen.rs:136]: just curious—why put all of them in `[]`? Does it simplify the resolution of the macro?

File: macro-rules/src/util.rs [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+        $($then)*
+    };
+
+    // Testing what kind of type we have (scalar vs ref)
+    //
+    // see java_types.rs
+
+    (if is_ref_ty($t:ident) { $($then:tt)* }) => {};
+
+    (if is_ref_ty($t:tt) { $($then:tt)* }) => {
```
Comment (rcoh) [inline macro-rules/src/util.rs:21]: this works because idents are never references but token trees are?

File: src/lib.rs [hunk 7]
```diff
@@ -68,6 +68,12 @@ pub mod plumbing {
     pub use crate::raw::{EnvPtr, FieldPtr, FromJniValue, IntoJniValue, MethodPtr, ObjectPtr};
     pub use crate::refs::NullJRef;
     pub use crate::to_java::{ToJavaImpl, ToJavaScalar};
+    pub use duchess_macro_rules::{
```
Comment (rcoh) [inline src/lib.rs:71]: I wonder if we should make the whole plumbing module doc(hidden)

**Review by rcoh — APPROVED** [review]
This is definitely a big improvement in terms of readability of the codegen.

In the long term, having some tests of the macros directly is probably a reasonable thing to add—I think it will make working on the macros easier.

There is some dead (and I think? incorrect) code in `macro_if` that we should remove before merging. But otherwise this looks good.

File: macro-rules/src/util.rs (hunk 3 above)
Comment (rcoh) [reply macro-rules/src/util.rs:9]: ah—I think actually all of these expect the last two are dead code

**Comment by nikomatsakis** [comment]
I'm of mixed minds about unit tests on the macros themselves — they don't represent a stable API boundary and I'm generally not in favor of tests that don't target a stable API boundary, since they can become dated so easily when refactorings occur. On the other hand, I know they can have value in making it easy to specify edge cases.

File: src/lib.rs (hunk 7 above)
Comment (nikomatsakis) [reply src/lib.rs:71]: Maybe. I've been thinking we should also rename it to `semver_unstable` or something more explicit.

File: macro-rules/src/lib.rs (hunk 2 above)
Comment (nikomatsakis) [reply macro-rules/src/lib.rs:1]: No, no way around that, but we could break them into a file per macro.

File: duchess-reflect/src/codegen.rs (hunk 5 above)
Comment (nikomatsakis) [reply duchess-reflect/src/codegen.rs:136]: It's not always necessary, but it is needed in some cases where a `,` is not a suitable "end delimiter", e.g. `$($t:tt)*` but also on occasion some other fragment specifiers. I used to only use it sometimes but I've wound up just adopting a consistent style of always including them.

File: macro-rules/src/util.rs (hunk 6 above)
Comment (nikomatsakis) [reply macro-rules/src/util.rs:21]: Yes, it's part of the invariant for how we represent types -- an identifier is a scalar or void, a `()` grouping is a reference. I could replace `$t:tt` with `($(t:tt)*)`, that might be better.

**Comment by nikomatsakis** [comment]
Thanks for the review, @rcoh, I'll go over the docs and make a few improvements and then merge.

**Comment by nikomatsakis** [comment]
Made cosmetic improvements. Will merge once CI is green.
